airflow/models/dagrun.py

```diff
@@ -286,25 +321,27 @@ def update_state(self, session=None):
             session=session
```

Comment:
Can you check if double calling get_task_instances is faster than filtering the list in Python? Line 319 and 306 contains calls to the get_task_instances method, and this method invokes a database query.

Reply:
Will check. I was mostly making these changes based on the hot-spots identified by the profling, but I have already extended this a bit beyond just that with the refactor of this method.

Perhaps I should revert some of these changes and make them in to a separate PR?

Reply:
Scheduler's performance is also an area of interest for Polidea and our clients, so dividing work into smaller PRs will allow us to develop better solutions. together